src/interface/meshblock_data.hpp

```diff
@@ -122,18 +115,27 @@ class MeshBlockData {
     pmy_block = other->GetBlockSharedPointer();
   }
 
-  void Initialize(const std::shared_ptr<StateDescriptor> resolved_packages,
-                  const std::shared_ptr<MeshBlock> pmb);
-
   /// Create copy of MeshBlockData, possibly with a subset of named fields,
   /// and possibly shallow.  Note when shallow=false, new storage is allocated
   /// for non-OneCopy vars, but the data from src is not actually deep copied
-  template <typename ID_t>
-  void Initialize(const MeshBlockData<T> *src, const std::vector<ID_t> &vars,
-                  const bool shallow_copy) {
+  template <class SRC_t, typename ID_t = std::string>
+  void Initialize(const std::shared_ptr<SRC_t> src, const std::vector<ID_t> &vars = {},
+                  const bool shallow_copy = false) {
+    Initialize(src->resolved_packages, src, vars, shallow_copy);
+  }
+
+  template <class SRC_t, typename ID_t = std::string>
+  void Initialize(const std::shared_ptr<StateDescriptor> resolved_packages_in,
+                  const std::shared_ptr<SRC_t> src, const std::vector<ID_t> &vars = {},
+                  const bool shallow_copy = false) {
+    if constexpr (!(std::is_same_v<SRC_t, MeshBlockData<Real>> ||
+                    std::is_same_v<SRC_t, MeshBlock>)) {
+      // We don't allow other types
+      static_assert(always_false<SRC_t>);
```

Comment:
Is there a reason not to just `static_assert` the boolean expression that you have in `if constexpr` here? Also you can add a string to the `static_assert` which I think is a nice readability enhancement.

Reply:
Yeah, there is some reason that `static_assert(false)` will always trigger a compiler failure even if that constexpr if branch isn't triggered. Having something that always evaluates to false that has a template argument does not get triggered in contrast. Apparently they fix this stupidity in c++23.

Reply:
added string to assert.